File: Problems/LongestPalindrome/longest_palindrome.rb
class Solution
  def longest_palindrome(s)
    length_of_string = s.length
    is_palindrome = Array.new(length_of_string) { Array.new(length_of_string, false) }

    start_of_string = 0
    end_of_string = 0

    (length_of_string - 1).downto(0) do |i|
      (i + 1).upto(length_of_string - 1) do |j|
        is_palindrome[i][j] = s[i] == s[j] && (j - i <= 2 || is_palindrome[i + 1][j - 1])
        if is_palindrome[i][j] && (j - i > end_of_string - start_of_string)
          start_of_string = i
          end_of_string = j
        end
      end
    end

    s[start_of_string..end_of_string]
  end
end

File: Problems/LongestPalindrome/longest_palindrome.py
def longest_palindrome(s: str) -> str:
    length_of_string = len(s)
    is_palindrome = [[False] * length_of_string for _ in range(length_of_string)]

    start_of_string = 0
    end_of_string = 0

    for i in range(length_of_string - 1, -1, -1):
        for j in range(i + 1, length_of_string):
            is_palindrome[i][j] = s[i] == s[j] and (j - i <= 2 or is_palindrome[i + 1][j - 1])
            if is_palindrome[i][j] and (j - i > end_of_string - start_of_string):
                start_of_string = i
                end_of_string = j

    return s[start_of_string:end_of_string + 1]
